keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,305 @@
+
+This proposal presents a solution for that problem.
+
+[#75294]: https://github.com/kubernetes/kubernetes/issues/75294
+
+### Goals
+
+- Reduce load on kube-apiserver and number of memory allocations, by avoiding
+serializing the same object multiple times for different watchers.
+
+### Non-Goals
+
+- Change overall architecture of the system, by changing what data is being
+read/watched by different components.
+
+## Proposal
+
+This proposal does not introduce any user-visible changes - the proposed changes
+are purely implementation details of kube-apiserver.
+
+The first observation is that a given object may be serialized to multiple
+different formats, based on it's:
+- group and version
+- subresource
+- output media type
+However, group, version and subresource are reflected in the `SelfLink` of
+returned object. As a result, for a given (potentially unversioned) object,
+we can identify all its possible serializations by (SelfLink, media-type)
+pairs, and that is what we will do below.
+
+We propose to extend [WithVersionEncoder][] by adding ObjectConvertor to it:
+```
+type WithVersionEncoder struct {
+	Version GroupVersioner
+	Encoder
+	ObjectConvertor
+	ObjectTyper
+}
+```
+
+On top of that, we propose introducing `CustomEncoder` interface:
+```
+type CustomEncoder interface {
+	InterceptEncode(encoder WithVersionEncoder, w io.Writer) error
+}
+```
+
+With that, we will change existing serializers (json, protobuf and versioning),
+to check if the to-be-serialized object implements that interface and if so simply
+call its `Encode()` method instead of using existing logic.
+
+[WithVersionEncoder]: https://github.com/kubernetes/kubernetes/blob/990ee3c09c0104cc1045b343040fe76082862d73/staging/src/k8s.io/apimachinery/pkg/runtime/helper.go#L215
+
+With those (very local and small) changes, we will introduce an internal type
+in package `cacher` implementing both `runtime.Object` and `CustomEncoder`
+interfaces. The idea behind it is that it will be encapsulating the original
+object and additionally it will be able to accumulate its serialized versions.
+It will look like this:
+```
+// TODO: Better name is welcome :)
+type CachingObject struct {
+	// Object is the object (potentially in the internal version)
+	// for which serializations should be cached for future reuse.
+	Object runtime.Object
+
+	// FIXME: We may want to change that during performance experiments
+	// e.g. to use sync.Map or slice or something different, also to
+	// allow some fast-path.
+	lock sync.Mutex
+	Versioned map[runtime.GroupVersioner]*CachingVersionedObject
+}
+
+// TODO: Better name is welcome :)
+type CachingVersionedObject struct {
+	// Ojbect is the versioned object for which serializations
+	// should be cached.
+
+	// FIXME: We may want to change that during performance experiments
+	// e.g. to use sync.Map or slice or something different, also to
+	// allow some fast-path.
+	lock sync.Mutex
+	serialization map[cachingKey]*serializationResult
```

Comment:
Given that (at least initially) we will use that only for watch (which doesn't support subresources), cachingKey can probably be simply replaced by encoder...